src/pgtz/src/timezone.rs

```diff
@@ -0,0 +1,497 @@
+
+/// Parsed timezone.
+#[derive(Arbitrary, Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, MzReflect)]
+pub enum Timezone {
+    #[serde(with = "fixed_offset_serde")]
+    FixedOffset(#[proptest(strategy = "any_fixed_offset()")] FixedOffset),
+    Tz(#[proptest(strategy = "any_timezone()")] Tz),
+}
+
+impl Timezone {
+    pub fn parse(tz: &str, spec: TimezoneSpec) -> Result<Self, String> {
+        build_timezone_offset_second(&tokenize_timezone(tz)?, tz, spec)
+    }
+}
+
+impl RustType<ProtoTimezone> for Timezone {
+    fn into_proto(&self) -> ProtoTimezone {
+        use proto_timezone::Kind;
+        ProtoTimezone {
+            kind: Some(match self {
+                Timezone::FixedOffset(fo) => Kind::FixedOffset(fo.into_proto()),
+                Timezone::Tz(tz) => Kind::Tz(tz.into_proto()),
+            }),
+        }
+    }
+
+    fn from_proto(proto: ProtoTimezone) -> Result<Self, TryFromProtoError> {
+        use proto_timezone::Kind;
+        let kind = proto
+            .kind
+            .ok_or_else(|| TryFromProtoError::missing_field("ProtoTimezone::kind"))?;
+        Ok(match kind {
+            Kind::FixedOffset(pof) => Timezone::FixedOffset(FixedOffset::from_proto(pof)?),
+            Kind::Tz(ptz) => Timezone::Tz(Tz::from_proto(ptz)?),
+        })
+    }
+}
+
+// We need to implement Serialize and Deserialize traits to include Timezone in the UnaryFunc enum.
+// FixedOffset doesn't implement these, even with the "serde" feature enabled.
+mod fixed_offset_serde {
+    use serde::de::Error;
+    use serde::{Deserializer, Serializer};
+
+    use super::*;
+
+    pub fn deserialize<'de, D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<FixedOffset, D::Error> {
+        let offset = i32::deserialize(deserializer)?;
+        FixedOffset::east_opt(offset).ok_or_else(|| {
+            Error::custom(format!("Invalid timezone offset: |{}| >= 86_400", offset))
+        })
+    }
+
+    pub fn serialize<S: Serializer>(
+        offset: &FixedOffset,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        serializer.serialize_i32(offset.local_minus_utc())
+    }
+}
+
+impl PartialOrd for Timezone {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// We need to implement Ord and PartialOrd to include Timezone in the UnaryFunc enum. Neither FixedOffset nor Tz
+// implement these so we do a simple ordinal comparison (FixedOffset variant < Tz variant), and break ties using
+// i32/str comparisons respectively.
+impl Ord for Timezone {
+    fn cmp(&self, other: &Self) -> Ordering {
+        use Timezone::*;
+        match (self, other) {
+            (FixedOffset(a), FixedOffset(b)) => a.local_minus_utc().cmp(&b.local_minus_utc()),
+            (Tz(a), Tz(b)) => a.name().cmp(b.name()),
+            (FixedOffset(_), Tz(_)) => Ordering::Less,
+            (Tz(_), FixedOffset(_)) => Ordering::Greater,
+        }
```

Comment:
Untested according to coverage report.

Reply:
This was code movement, not going to add a test at the moment.